Release notes for v0.34.0 (#2127)
grafana/k6: release notes/v0.34.0.md

k6 v0.34.0 is here! :tada: It introduces the already announced `k6/execution` API and includes some more enhancements and a bunch of minor bug fixes.

## New Features

* k6/execution API

Introduced a new module importable with the `k6/execution` path for fetching information about the execution state through a series of exported properties. It makes available information about VUs, Iterations and Scenarios.

> **Docs**: [grafana/xk6-execution](https://github.com/grafana/xk6-execution)

Reviewer: I would definitely expand on this with at least 1 example and what it's useful for

Author: I have added your gist in a simplified version.

## Enhancements and UX improvements

> * common.Bind has been replaced from a simpler and clearer API to get access to some useful information shared by registered modules. ([#2108](https://github.com/grafana/k6/pull/2108))

Reviewer: I would move this to internals and also probably link to the `modules` godoc

* Warn Windows users on importing dependencies as absolute paths. ([#2078](https://github.com/grafana/k6/pull/2078))

> * Speeding ups TC39 tests using a pool of Babel compilers. ([#1839](https://github.com/grafana/k6/pull/1839))

Reviewer: Again in internal changes

k6 v0.34.0 is here! :tada: It introduces the already announced `k6/execution` API and includes some more enhancements and a bunch of minor bug fixes.

## New k6 JavaScript API - `k6/execution`

> A new JS module with the `k6/execution` import was introduced. It allows the fetching of information about the current VU, instance (mostly relevant in distributed/cloud) or scenario execution state through a series of exported properties.

Reviewer: Given the changed heading I am for dropping of the whole first sentence
```suggestion
The `k6/execution` module allows the fetching of information about the current VU, instance (mostly relevant in distributed/cloud) or scenario execution state through a series of exported properties.
```
Maybe even `k6/execution` should link directly to it's JS API docs

Author: maintain it as unresolved so I will add the link when the doc will be merged

import { SharedArray } from "k6/data";

const data = new SharedArray("my dataset", function(){
	// generate and return the data
})

export const options = {
    scenarios :{
        "use-all-the-data": {
            executor: "shared-iterations",
            vus: 100,
            iterations: data.length,
            maxDuration: "1h"
        }
    }
}

export default function() {
    // this is unique even in the cloud
    var item = data[exec.scenario.iterationInTest];
    http.post("https://httpbin.test.k6.io/amazing-endpoint", item)
}
```

**Docs**: [grafana/xk6-execution](https://github.com/grafana/xk6-execution)

## Enhancements and UX improvements

* Warn Windows users on importing dependencies or opening files as absolute paths ([#2078](https://github.com/grafana/k6/pull/2078)).
* Pass setup data object into [handleSummary callback](https://k6.io/docs/results-visualization/end-of-test-summary/#handlesummary-callback) ([#2103](https://github.com/grafana/k6/pull/2103)). Thanks, @SamuelJohnson01997!

## Breaking changes

* The deprecated outputs Datadog and Kafka have been removed ([#2081](https://github.com/grafana/k6/pull/2081)).

## Bugs fixed!

> * Use HTTP POST method instead of GET for pushing logs requests to Loki ([#2100](https://github.com/grafana/k6/pull/2100)).

Reviewer 2: ```suggestion
* Use the `POST` HTTP request method instead of `GET` for pushing logs to Loki ([#2100](https://github.com/grafana/k6/pull/2100)).
```

> * Encode the BlacklistIP option using the CIDR notation ([#2083](https://github.com/grafana/k6/pull/2083)).

Reviewer 2: ```suggestion
* Encode the [`blacklistIPs` option](https://k6.io/docs/using-k6/options/#blacklist-ips) using the CIDR notation in JSON ([#2083](https://github.com/grafana/k6/pull/2083)).
```

* `ext.loadimpact` option has the same precedence as the script configuration during the consolidation process ([#2099](https://github.com/grafana/k6/pull/2099)).
* The WebSocket connection used for tailing logs from the k6 Cloud is reestablished in the case of an unexpected error ([#2090](https://github.com/grafana/k6/pull/2090)).

## Internals

> * A simpler and clearer API has been added as an alternative to `common.Bind`, which also gives JS modules easy access to some [useful internal objects and runtime information](https://pkg.go.dev/go.k6.io/k6@v0.34.0/js/modules#Instance) ([#2108](https://github.com/grafana/k6/pull/2108)).

Reviewer 2: ```suggestion
* A simpler and clearer API has been added as an alternative to `common.Bind`, which also gives JS modules and extensions easy access to some [useful internal objects and runtime information](https://pkg.go.dev/go.k6.io/k6@v0.34.0/js/modules#Instance) ([#2108](https://github.com/grafana/k6/pull/2108)). This API is not yet stable, it's very likely to change more in future k6 versions.
```

* Speeding ups TC39 tests using a pool of Babel compilers ([#1839](https://github.com/grafana/k6/pull/1839)).

> * Goja dependency has been updated adding the native support for Arrow functions, Destructuring, Default arguments and Computed properties features. For the same reason, the relative Babel's plugins supporting those features are not required anymore so they have been disabled ([#2109](https://github.com/grafana/k6/pull/2109)).

Reviewer 2: This should probably be in the `Enhancements and UX improvements` category, right? After all, we now support some destructuring modes that we didn't support before, even with Babel, right @MStoykov?

Reviewer: I guess we do now support https://github.com/grafana/k6/issues/824, but apart from that almost everything else is *very* minor so I wouldn't really qualify it as that much of an improvement, but we can move it

Reviewer 2: it's certainly not an entirely internal thing, if we now support some JS aspect we didn't before, so let's move at least that part

We can also keep a mention here that we've updated some dependencies and link both to the goja update PRs and to https://github.com/grafana/k6/pull/2092

Author: Updated. @MStoykov WDYT? Is it clear?

k6 v0.34.0 is here! :tada: It introduces the already announced `k6/execution` API and includes some more enhancements and a bunch of minor bug fixes.

## New k6 JavaScript API - `k6/execution`

The [k6/execution](https://k6.io/docs/javascript-api/k6-execution) module allows the fetching of information about the current VU, instance (mostly relevant in distributed/cloud) or scenario execution state through a series of exported properties.

### How to use

> Some of the previously suggested hacks with the globally available `__VU` and `__ITER` values, such as for [getting a unique object per iteration from an array](https://community.k6.io/t/when-parameterizing-data-how-do-i-not-use-the-same-data-more-than-once-in-a-test/42/2) or [`SharedArray`](https://k6.io/docs/javascript-api/k6-data/sharedarray/), can be done by using the `scenario.iterationInTest` property, which is guaranteed to be unique across VUs, even for distributed or cloud tests. For example:

Reviewer 3: "Hacks" sounds a bit too strong :)

```suggestion
Some of the previously suggested solutions with the globally available `__VU` and `__ITER` values, such as for [getting a unique object per iteration from an array](https://community.k6.io/t/when-parameterizing-data-how-do-i-not-use-the-same-data-more-than-once-in-a-test/42/2) or [`SharedArray`](https://k6.io/docs/javascript-api/k6-data/sharedarray/), can be done by using the `scenario.iterationInTest` property, which is guaranteed to be unique across VUs, even for distributed or cloud tests. For example:
```

> ```

Reviewer 2: ```suggestion
```js
```

import exec from "k6/execution";
import { SharedArray } from "k6/data";

const data = new SharedArray("my dataset", function(){

> 	return JSON.parse(open('my-large-dataset.json'));

Reviewer 2: ```suggestion
    return JSON.parse(open('my-large-dataset.json'));
```

})

export const options = {

>     scenarios :{
>         "use-all-the-data": {
>             executor: "shared-iterations",
>             vus: 100,
>             iterations: data.length,
>             maxDuration: "1h"
>         }
>     }
> }
> 
> export default function() {
>     // this is unique even in the cloud
>     var item = data[exec.scenario.iterationInTest];
>     http.post("https://httpbin.test.k6.io/amazing-endpoint", item)

Reviewer 3: ```suggestion
  scenarios :{
    "use-all-the-data": {
      executor: "shared-iterations",
      vus: 100,
      iterations: data.length,
      maxDuration: "1h"
    }
  }
}

export default function() {
  // this is unique even in the cloud
  var item = data[exec.scenario.iterationInTest];
  http.post("https://httpbin.test.k6.io/amazing-endpoint", item)
```

>     http.post("https://httpbin.test.k6.io/amazing-endpoint", item)

Reviewer 2: ```suggestion
   https://httpbin.test.k6.io/anything?endpoint=amazing
```

`amazing-endpoint` is not an actual valid url, returns 404

Reviewer: I think we still need the `http.post` ;)

}
```

> **Docs**: [k6/execution](https://k6.io/docs/javascript-api/k6-execution)

Reviewer 3: ```suggestion
You can read the full documentation [here](https://k6.io/docs/javascript-api/k6-execution).
```

Author: This is a copy-paste from the template. Should we change it also there?

## Enhancements and UX improvements

* Warn Windows users on importing dependencies or opening files as absolute paths ([#2078](https://github.com/grafana/k6/pull/2078)).
* Pass setup data object into [handleSummary callback](https://k6.io/docs/results-visualization/end-of-test-summary/#handlesummary-callback) ([#2103](https://github.com/grafana/k6/pull/2103)). Thanks, @SamuelJohnson01997!

> * Overwrite params using the spread operator ([#824](https://github.com/grafana/k6/issues/824), [#2109](https://github.com/grafana/k6/pull/2109)).

Reviewer 2: ```suggestion
```

Turns out this wasn't working as expected in the default compatibility mode because of some Babel issues, so we probably shouldn't mention it.

## Breaking changes

> * The deprecated outputs Datadog and Kafka have been removed ([#2081](https://github.com/grafana/k6/pull/2081)).

Reviewer 3: We should probably mention the new alternative if people were still using them.
```suggestion
* The deprecated outputs Datadog and Kafka have been removed ([#2081](https://github.com/grafana/k6/pull/2081)). The `statsd` output with environment variable `K6_STATSD_ENABLE_TAGS=true` can be used instead.
```

k6 v0.34.0 is here! :tada: It introduces the already announced `k6/execution` API and includes some more enhancements and a bunch of minor bug fixes.

> ## New Features
> 
> * k6/execution API

Reviewer 2: ```suggestion
## New k6 JavaScript API - `k6/execution`
```

This seems both more readable ([check out how the current one is rendered](https://github.com/grafana/k6/blob/release-notes-34/release%20notes/v0.34.0.md)) and more informative

> Introduced a new module importable with the `k6/execution` path for fetching information about the execution state through a series of exported properties. It makes available information about VUs, Iterations and Scenarios.

Reviewer 2: ```suggestion
A new JS module with the `k6/execution` import path was introduced. It allows the fetching of information about the current VU, iteration or scenario execution state through a series of exported properties.
```

> Introduced a new module importable with the `k6/execution` path for fetching information about the execution state through a series of exported properties. It makes available information about VUs, Iterations and Scenarios.

Author: Added but check the latest version because I had to integrate with https://github.com/grafana/k6/pull/2127#discussion_r703486009. Is it fine?

### How to use

> Accessing items iterating a dataset or getting them just once time cases are easier now. 

Author: This needs to be rearranged, if you have better words, please, share them.

> Accessing items iterating a dataset or getting them just once time cases are easier now. 

Reviewer 3: ```suggestion
Some of the previously suggested use cases for the global `__VU` and `__ITER` values, such as for [getting a unique object per iteration from an array](https://community.k6.io/t/when-parameterizing-data-how-do-i-not-use-the-same-data-more-than-once-in-a-test/42/2) or [`SharedArray`](https://k6.io/docs/javascript-api/k6-data/sharedarray/), can be simplified by using the `scenario.iterationInTest` property, which is guaranteed to be unique across VUs, even for tests in the k6 Cloud. For example:
```

> Some of the previously suggested use cases for the global `__VU` and `__ITER` values, such as for [getting a unique object per iteration from an array](https://community.k6.io/t/when-parameterizing-data-how-do-i-not-use-the-same-data-more-than-once-in-a-test/42/2) or [`SharedArray`](https://k6.io/docs/javascript-api/k6-data/sharedarray/), can be simplified by using the `scenario.iterationInTest` property, which is guaranteed to be unique across VUs, even for tests in the k6 Cloud. For example:

Reviewer: ```suggestion
Some of the previously suggested hacks with the globally available `__VU` and `__ITER` values, such as for [getting a unique object per iteration from an array](https://community.k6.io/t/when-parameterizing-data-how-do-i-not-use-the-same-data-more-than-once-in-a-test/42/2) or [`SharedArray`](https://k6.io/docs/javascript-api/k6-data/sharedarray/), can be done by using the `scenario.iterationInTest` property, which is guaranteed to be unique across VUs, even for distributed or cloud tests. For example:
```

```
import exec from "k6/execution";
import { SharedArray } from "k6/data";

const data = new SharedArray("my dataset", function(){

> 	// generate and return the data

Reviewer 2: ```suggestion
	return JSON.parse(open('my-large-dataset.json')); 
```
no need to over-complicate it

})

export const options = {
    scenarios :{
        "use-all-the-data": {
            executor: "shared-iterations",
            vus: 100,
            iterations: data.length,
            maxDuration: "1h"
        }
    }
}

export default function() {
    // this is unique even in the cloud
    var item = data[exec.scenario.iterationInTest];
    http.post("https://httpbin.test.k6.io/amazing-endpoint", item)
}
```

> **Docs**: [grafana/xk6-execution](https://github.com/grafana/xk6-execution)

Reviewer 2: We'll have to update this to actual docs URL after https://github.com/grafana/k6-docs/pull/396

## Enhancements and UX improvements

> * Warn Windows users on importing dependencies as absolute paths. ([#2078](https://github.com/grafana/k6/pull/2078))

Reviewer: ```suggestion
* Warn Windows users on importing dependencies or opening files as absolute paths. ([#2078](https://github.com/grafana/k6/pull/2078))
```

* Pass setup data object into [handleSummary callback](https://k6.io/docs/results-visualization/end-of-test-summary/#handlesummary-callback). ([#2103](https://github.com/grafana/k6/pull/2103)) - thanks @SamuelJohnson01997

## Breaking changes

* The deprecated outputs Datadog and Kafka have been removed. ([#2081](https://github.com/grafana/k6/pull/2081))

## Bugs fixed!

* Use HTTP POST method instead of GET for pushing logs requests to Loki. ([#2100](https://github.com/grafana/k6/pull/2100))
* Encode the BlacklistIP option using the CIDR notation. ([#2083](https://github.com/grafana/k6/pull/2083))
* `ext.loadimpact` option has the same precedence as the script configuration during the consolidation process. ([#2099](https://github.com/grafana/k6/pull/2099))

> * Log tailing from cloud refreshes the unedrhood websocket connection in the case an unexpected error occurred. ([#2090](https://github.com/grafana/k6/pull/2090))

Reviewer 3: ```suggestion
* The WebSocket connection used for tailing logs from the k6 Cloud is reestablished in the case of an unexpected error. ([#2090](https://github.com/grafana/k6/pull/2090))
```

## Internals

> * common.Bind has been replaced from a simpler and clearer API to get access to some useful information shared by registered [modules](https://pkg.go.dev/go.k6.io/k6@v0.33.1-0.20210831121908-a96a8abf4bb5/js/modules#Instance). ([#2108](https://github.com/grafana/k6/pull/2108))

Reviewer 3: It wasn't replaced (removed) yet, so maybe:

```suggestion
* A simpler and clearer API has been added as an alternative to `common.Bind`, which also gives JS modules easy access to some [useful internal objects and runtime information](https://pkg.go.dev/go.k6.io/k6@v0.34.0/js/modules#Instance). ([#2108](https://github.com/grafana/k6/pull/2108))
```

Also maybe link to the final `v0.34.0` URL, though it will 404 until `pkg.go.dev` is updated. So maybe not? :man_shrugging:

Author: Keeping this as unresolved for maintaining the working link available https://pkg.go.dev/go.k6.io/k6@v0.33.1-0.20210831121908-a96a8abf4bb5/js/modules#Instance
